### DISCUSSION TITLE
add topo to mesh WORLD

### DISCUSSION THREAD

**Discussion by cecidip** [gimli-org/pyGIMLi#645] [Q&A] [answered]
I'm using `mesh = pg.meshtools.createMesh(world, quality=33, area=40)`, since my acquisition had geophones very very close to each other, therefore `mesh = mgr.createMesh(data)` has too many celle numbers, and the inversion is not only slow but bad.
Anyways, i would like to add my topography to the mesh, since it is very considerable, how can i do it ? 
This is the result i get with the non-topo mesh (rays and final result) :
![image](https://github.com/gimli-org/gimli/assets/22650615/bc3836e0-76bf-4f1d-b224-597ce7cab3db)

![image](https://github.com/gimli-org/gimli/assets/22650615/c0473d2a-81e9-4528-9342-6b3e2fc3d22b)

Thank you !

Extra info: 
-Z information is contained in the data container at `data.setSensors(pos)`
- this is the mesh created with world: 
`Mesh: Nodes: 3685 Cells: 6875 Boundaries: 10559
`![image](https://github.com/gimli-org/gimli/assets/22650615/ab88aaa2-6c39-4271-8c8b-aecda0448822)
- this is the mesh created with  mgr.createMesh:
`Mesh: Nodes: 228300 Cells: 407248 Boundaries: 635547
`![image](https://github.com/gimli-org/gimli/assets/22650615/de68d16f-eb8e-4611-a284-7f8cf18fe068)

**Comment by halbmy** [comment, marked as answer]
We refer to: https://www.pygimli.org/pygimliapi/_generated/pygimli.meshtools.html 
`createWorld` is, as the short summary in says for 
> Create simple rectangular 2D or 3D world
For generating a mesh using the topography in a data file, one should use `createParaMesh(data)`, which in turn calls `createParaMeshPLC` and then `createMesh`, e.g.:
```
import pygimli.meshtools as mt
plc = mt.createParaMeshPLC(data, boundary=0, paraDX=0.5, paraDepth=100, area=40)
mesh = mt.createMesh(plc, quality=30, )
```
The first takes refinement at sensors, depth, maximum cell size, boundary size etc. (see docstring), and the latter the mesh quality. One can add any objects like lines or anomalies to plc before meshing. Have a look at the docstrings, tutorials and examples.

**Reply by cecidip** [reply]
Thank you, but as i mentioned, if i use the sensors for my mesh refinement it does not work, cause they are too close. Also, with that many number of cells my kerner keeps crashing. So, I'd like to know if there is any way to create a mesh following the topohraphy, but not the sensors refinement. 
Running the lines you suggested this is the mesh info:
`Mesh: Nodes: 236478 Cells: 406186 Boundaries: 642663`
Even if i change de depth or area, it satays in that order of magnitude. This is the result, and the zoomed version  so you can see the problem of refining by sensor.
![Screenshot 2024-02-02 at 1 14 32 PM](https://github.com/gimli-org/gimli/assets/22650615/72d16326-61fb-43b0-98cb-16487061faa9)

![Screenshot 2024-02-02 at 1 14 18 PM](https://github.com/gimli-org/gimli/assets/22650615/9850a1ad-d168-4568-b1cf-2f24f7cc8e5e)

I found that World uses a reasonable refinement ( and number of cells in consequence) but since it does not include the topo, the inversion is bad. Any suggestions ?

**Comment by halbmy** [comment]
Do your sensors have to be nodes in the mesh (recommended at least for ERT)? If not, you can just interpolate your topography on a linear x spacing and create the PLC from there. If yes, you can avoid refinement by `paraDX=0`. Otherwise you can try to snap very close positions to common points.

**Reply by cecidip** [reply]
No, they dont need to be nodes. That's exacty what i just did, created a topo repository and created a mesh with it. The mesh and cell number look good ! Let's hope the inversion works. Thank you
